### PR TITLE
[SPARK-39339][SQL] Support TimestampNTZ type in JDBC data source

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

This PR adds support for TimestampNTZ (TIMESTAMP WITHOUT TIME ZONE) in JDBC data source. It also introduces a new configuration option `inferTimestampNTZType` which allows to read written timestamps as timestamp without time zone. By default this is set to `false`, i.e. all timestamps are read as legacy timestamp type.

Here is the state of timestamp without time zone support in the built-in dialects:
- H2: timestamp without time zone, seems to map to timestamp type
- Derby: only has timestamp type
- MySQL: only has timestamp type
- Postgres: has timestamp without time zone, which maps to timestamp
- SQL Server: only datetime/datetime2, neither are time zone aware
- Oracle: seems to only have timestamp and timestamp with time zone
- Teradata: similar to Oracle but I could not verify
- DB2: has TIMESTAMP WITHOUT TIME ZONE but I could not make this type work in my test, only TIMESTAMP seems to work

### Why are the changes needed?

Adds support for the new TimestampNTZ type, see https://issues.apache.org/jira/browse/SPARK-35662.

### Does this PR introduce _any_ user-facing change?

JDBC data source is now capable of writing and reading TimestampNTZ types. When reading timestamp values, configuration option `inferTimestampNTZType` allows to infer those values as TIMESTAMP WITHOUT TIME ZONE. By default the option is set to `false` so the behaviour is unchanged and all timestamps are read TIMESTAMP WITH LOCAL TIME ZONE.

### How was this patch tested?

I added a unit test to ensure the general functionality works. I also manually verified the write/read test for TimestampNTZ in the following databases (all I could get access to):
- H2, `jdbc:h2:mem:testdb0`
- Derby, `jdbc:derby:<filepath>`
- MySQL, `docker run --name mysql -e MYSQL_ROOT_PASSWORD=secret -e MYSQL_DATABASE=db -e MYSQL_USER=user -e MYSQL_PASSWORD=secret -p 3306:3306 -d mysql:5.7`, `jdbc:mysql://127.0.0.1:3306/db?user=user&password=secret`
- PostgreSQL, `docker run -d --name postgres -e POSTGRES_PASSWORD=secret -e POSTGRES_USER=user -e POSTGRES_DB=db -p 5432:5432 postgres:12.11`, `jdbc:postgresql://127.0.0.1:5432/db?user=user&password=secret`
- SQL Server, `docker run -e "ACCEPT_EULA=Y" -e SA_PASSWORD='yourStrong(!)Password' -p 1433:1433 -d mcr.microsoft.com/mssql/server:2019-CU15-ubuntu-20.04`, `jdbc:sqlserver://127.0.0.1:1433;user=sa;password=yourStrong(!)Password`
- DB2, ` docker run -itd --name mydb2 --privileged=true -p 50000:50000 -e LICENSE=accept -e DB2INST1_PASSWORD=secret -e DBNAME=db ibmcom/db2`, `jdbc:db2://127.0.0.1:50000/db:user=db2inst1;password=secret;`.

### CODE REVIEW

**Comment by sadikovi** [comment]
@gengliangwang @beliefer Can you review this PR? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 1]
```diff
@@ -226,6 +226,9 @@ class JDBCOptions(
   // The prefix that is added to the query sent to the JDBC database.
   // This is required to support some complex queries with some JDBC databases.
   val prepareQuery = parameters.get(JDBC_PREPARE_QUERY).map(_ + " ").getOrElse("")
+
+  // Infers timestamp values as TimestampNTZ type when reading data.
+  val inferTimestampNTZType = parameters.getOrElse(JDBC_INFER_TIMESTAMP_NTZ, "false").toBoolean
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:231]: Should we maybe check if `spark.sql.timestampType` is `TIMESTAMP_NTZ` if `inferTimestampNTZType` is not set? That's how CSV type inference and Python type inference do.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:231]: cc @gengliangwang FYI

**Comment by 1104056452** [comment]
Can someone help review this patch？thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 1 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:231]: Yes, I thought about it, let's ask @gengliangwang.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -472,6 +481,15 @@ object JdbcUtils extends Logging with SQLConfHelper {
...
         }
 
+    case TimestampNTZType =>
+      (rs: ResultSet, row: InternalRow, pos: Int) =>
+        val t = rs.getTimestamp(pos + 1)
+        if (t != null) {
+          row.setLong(pos, DateTimeUtils.fromJavaTimestamp(t))
+        } else {
+          row.update(pos, null)
+        }
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:491]: Seems same as `TimestampType` branch, should we merge them?

https://github.com/apache/spark/blob/8bbbdb5a4667d90b27b202870cd73c9a19b74781/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L466-L473

**Comment by sadikovi** [comment]
@beliefer Can you review this PR from JDBC perspective? I think you have contributed extensively to this part of the code. Also, cc @gengliangwang.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 3]
```diff
@@ -1879,5 +1880,27 @@ class JDBCSuite extends QueryTest
     val fields = schema.fields
     assert(fields.length === 1)
     assert(fields(0).dataType === StringType)
-   }
+  }
+
+  test("SPARK-39339: TimestampNTZType support") {
+    val tableName = "timestamp_ntz_table"
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1886]: Could you add tests write/read timestamp_ntz with different time zone ? I doubt the result is not correct.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 4]
```diff
@@ -150,6 +150,9 @@ object JdbcUtils extends Logging with SQLConfHelper {
       case StringType => Option(JdbcType("TEXT", java.sql.Types.CLOB))
       case BinaryType => Option(JdbcType("BLOB", java.sql.Types.BLOB))
       case TimestampType => Option(JdbcType("TIMESTAMP", java.sql.Types.TIMESTAMP))
+      // Most of the databases either don't support TIMESTAMP WITHOUT TIME ZONE or map it to
+      // TIMESTAMP type. This will be overwritten in dialects.
+      case TimestampNTZType => Option(JdbcType("TIMESTAMP", java.sql.Types.TIMESTAMP))
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:155]: We cannot do this. We should let JDBC dialect decide how to do the mapping.

Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:155]: This is a generic case. JDBC dialects can override this setting if they want to. I confirm that most of them work fine with TIMESTAMP.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 3 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1886]: Timestamp NTZ type is independent of the time zone. Sure, I can add a test to confirm.

**Comment by sadikovi** [comment]
@gengliangwang Can you review?

Thanks to the comments from the reviewers, I noticed that there could be inconsistencies when writing timestamp with local time zone and reading as timestamp_ntz. For example, writing 2020-01-01 00:00:00 in local time zone (UTC+1), timestamp_ntz will be read as 2019-12-31 23:00:00. This is because we always store timestamp in UTC and we don't have information on what time zone was used when writing timestamp.

Could you advise on how to proceed? I am not sure we can do much about it because we don't store time zone information.

**Comment by gengliangwang** [comment]
@sadikovi yes will do. I just moved home. Sorry for the late reply.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 5]
```diff
@@ -1879,5 +1880,53 @@ class JDBCSuite extends QueryTest
...
+            .mode("overwrite")
+            .option("url", urlWithUserAndPass)
+            .option("dbtable", tableName)
+            .save()
+
+          val res = spark.read.format("jdbc")
+            .option("inferTimestampNTZType", "true")
+            .option("url", urlWithUserAndPass)
+            .option("dbtable", tableName)
+            .load()
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1925]: This test case always read/write with the same time zone. You can reference https://github.com/apache/spark/blob/e410d98f57750080ad46932cc9211d2cf5154c24/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcQuerySuite.scala#L808

Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1925]: Yes, I will update, thanks 👍.

**Review by gengliangwang — APPROVED** [review]
LGTM, thanks for the work!

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1925]: The test case still read and write to JDBC with the same time zone.

**Comment by sadikovi** [comment]
@gengliangwang I made a few small changes. Can you review again? Thanks.

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by beliefer** [comment]
> Thanks, merging to master

I update this test case and it will fail !
```
  test("SPARK-37463: read/write Timestamp ntz to Orc with different time zone") {
    DateTimeTestUtils.withDefaultTimeZone(DateTimeTestUtils.LA) {
      val sqlText = """
                      |select
                      | timestamp_ntz '2021-06-01 00:00:00' ts_ntz1,
                      | timestamp_ntz '1883-11-16 00:00:00.0' as ts_ntz2,
                      | timestamp_ntz '2021-03-14 02:15:00.0' as ts_ntz3
                      |""".stripMargin

      withTempPath { dir =>
        val path = dir.getCanonicalPath
        val df = sql(sqlText)

        df.write.mode("overwrite").orc(path)

        val query = s"select * from `orc`.`$path`"

        DateTimeTestUtils.outstandingZoneIds.foreach { zoneId =>
          DateTimeTestUtils.withDefaultTimeZone(zoneId) {
            withAllNativeOrcReaders {
              checkAnswer(sql(query), df)
            }
          }
        }
      }
    }
  }
```

**Comment by gengliangwang** [comment]
@beliefer Your test case is testing ORC, while this PR is about JDBC...

**Comment by sadikovi** [comment]
I updated the test case as you suggested and it passes on my machine. Can you share the error message? It also passed the build.

**Comment by beliefer** [comment]
I think we can't support timestamp ntz with the option. 
We should let JDBC dialect to decide how to supports timestamp ntz.
If one table have ts1 is timestamp and ts2 is timestamp ntz, what is the output when we specify the `inferTimestampNTZType` option ?

**Comment by beliefer** [comment]
> I updated the test case as you suggested and it passes on my machine. Can you share the error message? It also passed the build.

```
== Results ==
!== Correct Answer - 1 ==                                           == Spark Answer - 1 ==
 struct<TIMESTAMP_NTZ '1500-01-20 00:00:00.123456':timestamp_ntz>   struct<TIMESTAMP_NTZ '1500-01-20 00:00:00.123456':timestamp_ntz>
![1500-01-20T00:00:00.123456]                                       [1500-01-20T00:16:08.123456]
```

**Comment by sadikovi** [comment]
I think we can. JDBC dialects can configure how they map TimestampNTZ type.
In the case you mentioned, both timestamps will be read as timestamp_ntz in MySQL and Postgres. In fact, the current timestamp type is stored as timestamp_ntz in those database systems.

Even with dialects managing timestamp_ntz writes and reads, this would be the same problem unless you store them as different types.

Also, the test passes in master:
```
[ivan.sadikov@C02DV1TGMD6R spark-oss (master)]$ git log -n1
commit 2349175e1b81b0a61e1ed90c2d051c01cf78de9b (HEAD -> master, upstream/master)
Author: Ivan Sadikov <ivan.sadikov@databricks.com>
Date:   Mon Jun 13 21:22:15 2022 -0700

    [SPARK-39339][SQL] Support TimestampNTZ type in JDBC data source
    



[info] JDBCSuite:
05:53:04.233 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
05:53:07.936 ERROR org.apache.spark.sql.execution.datasources.jdbc.connection.ConnectionProvider: Failed to load built-in provider.
[info] - SPARK-39339: Handle TimestampNTZType null values (1 second, 555 milliseconds)
[info] - SPARK-39339: TimestampNTZType with different local time zones (4 seconds, 48 milliseconds)
05:53:14.022 WARN org.apache.spark.sql.jdbc.JDBCSuite: 

===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.jdbc.JDBCSuite, threads: Timer-2 (daemon=true), rpc-boss-3-1 (daemon=true), shuffle-boss-6-1 (daemon=true) =====
[info] Run completed in 11 seconds, 724 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 2, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 81 s (01:21), completed Jun 14, 2022 5:53:14 AM
```

**Comment by sadikovi** [comment]
@beliefer Maybe we can address your concerns in the follow-up work, what do you think? We can open a follow-up ticket and try to polish the implementation - it is not perfect by any means!

**Comment by beliefer** [comment]
@sadikovi You can run the test case I added above.

**Pull request closed without merging**